core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java

```diff
@@ -1371,24 +1371,36 @@ public AlluxioURI translateUri(String uriStr) throws InvalidPathException {
     }
 
     // Reverse lookup mount table to find mount point that contains the path.
-    try {
-      return mMountTable.reverseLookup(uriStr);
-    } catch (InvalidPathException e) {
-      // Throw if auto-mount is not requested.
-      if (!ServerConfiguration.getBoolean(PropertyKey.MASTER_SHIMFS_AUTO_MOUNT_ENABLED)) {
-        throw e;
+    MountTable.ReverseResolution resolution = mMountTable.reverseResolve(uri);
+    if (resolution != null) {
+      // Before returning the resolved path, make sure mount point does not contain
+      // nested mounts as translation on one mount should not give visibility to another.
+      if (mMountTable.containsMountPoint(resolution.getMountInfo().getAlluxioUri(), false)) {
+        throw new IllegalStateException(
+            String.format("Foreign URI: %s is in a mount that has nested mounts.", uriStr));
```

Comment:
can we include both the ufsUri and the reversed resolved alluxioUri in the exception message?